buildflow/core/credentials/clickhouse_credentials.py

```diff
@@ -0,0 +1,11 @@
+from buildflow.core.credentials._credentials import Credentials
+from buildflow.core.options.credentials_options import CredentialsOptions
+
+
+class ClickhouseCredentials(Credentials):
```

Comment:
I would lean to these being arguments to the primitive itself similar to how we do for snowflake: https://github.com/launchflow/buildflow/blob/78ba132a775faca6537e90eacbeb8712edc89a64/buildflow/io/snowflake/snowflake_table.py#L37C1-L37C1

Our general rule of thumb is to keep them there unless we have multiple primitives using the credentials (like AWS and GCP).

Reply:
For the snowflake sink case, it is mentioned the [type](https://github.com/launchflow/buildflow/blob/78ba132a775faca6537e90eacbeb8712edc89a64/buildflow/io/snowflake/snowflake_table.py#L153) of credentials would be either AWSCredentails to GCPCredentials (I am not sure whether AWS or GCPCredentials are passed to use them as a data source for snowflake). But for the clickhouse case can I mention the type as a `dict` as simple as that which contains the credentials that can be passed to the `ClickhouseSink`?

Reply:
Snowflake is a little special because it uses a GCP or AWS bucket for ingesting data into snowflake. Then their are additional fields for the actual authentication with snowflake (private key, user, etc..)

I think for clickhouse sink I would just add three new fields that can be passed in:
- host
- username
- password

Reply:
@boetro But, the `Strategy` (inherited by the `ClickhouseStrategies`) expects a `CredentialType`, whereas the `Credentials` class expects a `CredentialsOptions` which is nothing but a dataclass like `AWSCredentials`. That's why I created a class for Clickhouse (`ClickhouseCredentials`). But you said to remove because it's not needed by multiple primitives. So we can't pass the host, username, and password.

Reply:
A dataclass can be created using `dataclasses.make_dataclass` too like this:

```
Credentials = make_dataclass('Credentials',
                                     [('host', str, field(default='localhost')),
                                      ('username', str, field(default='username')),
                                      ('password', str, field(default='password'))])
credentials = Credentials(host, username, password)
super().__init__(credentials=credentials, strategy_id="clickhouse-sink")
```

The above code will go into the constructor of `ClickhouseStrategies`. Here's the complete code for the context,

```
class ClickhouseSink(SinkStrategy):
    def __init__(
        self,
        *,
        host: ClickhouseHost,
        username: ClickhouseUsername,
        password: ClickhousePassword,
        database: ClickhouseDatabase,
        table: ClickhouseTableID,
    ):
        Credentials = make_dataclass('Credentials',
                                     [('host', str, field(default='localhost')),
                                      ('username', str, field(default='username')),
                                      ('password', str, field(default='password'))])
        credentials = Credentials(host, username, password)
        super().__init__(credentials=credentials, strategy_id="clickhouse-sink")
        self.client = None
        self.database = database
        self.table = table
        self.connect(credentials)
```

Since the expectation for the credentials is the dataclass by the `Strategy` class we can tackle the problem like this!

Reply:
I think that all makes sense to me. Another option would be to "ignore" the super classes credentials and just pass in the [EmptyCredentials](https://github.com/launchflow/buildflow/blob/cb7001c731d41f22ddf491f123b037dfd5b7fa35/buildflow/core/credentials/empty_credentials.py#L4) class. That's how the [duckdb sink does things.](https://github.com/launchflow/buildflow/blob/cb7001c731d41f22ddf491f123b037dfd5b7fa35/buildflow/io/duckdb/strategies/duckdb_strategies.py#L24) but I think I prefer your proposal!